src/recipe/stage2.rs

```diff
@@ -0,0 +1,1918 @@
+        Self::from_raw(raw, jinja_opt).map_err(|err| ParsingError::from_partial(yaml, err))
+    }
+
+    /// Build a recipe from a YAML string and use a given package hash string as default value.
+    pub fn from_yaml_with_default_hash_str(
+        yaml: &str,
+        default_pkg_hash: &str,
+        jinja_opt: SelectorConfig,
+    ) -> Result<Self, ParsingError> {
+        let mut recipe = Self::from_yaml(yaml, jinja_opt)?;
+
+        // Set the build string to the package hash if it is not set
+        if recipe.build.string.is_none() {
+            recipe.build.string = Some(format!("{}_{}", default_pkg_hash, recipe.build.number));
+        }
+        Ok(recipe)
+    }
+
+    /// Build a recipe from a [`RawRecipe`].
+    pub fn from_raw(
+        raw: RawRecipe,
+        jinja_opt: SelectorConfig,
+    ) -> Result<Self, PartialParsingError> {
+        // Init minijinja
+        let mut jinja = Jinja::new(jinja_opt);
+
+        for (k, v) in raw.context {
+            let rendered = jinja.render_str(v.as_str()).map_err(|err| {
+                _partialerror!(
+                    *v.span(),
+                    ErrorKind::JinjaRendering(err),
+                    label = "error rendering context"
+                )
+            })?;
+
+            jinja
+                .context_mut()
+                .insert(k.as_str().to_owned(), Value::from_safe_string(rendered));
+        }
+
+        let package = Package::from_stage1(&raw.package, &jinja)?;
+        let source = Source::from_stage1(raw.source, &jinja)?;
+
+        let about = raw
+            .about
+            .as_ref()
+            .map(|about| About::from_stage1(about, &jinja))
+            .transpose()?
+            .unwrap_or_default();
+
+        let requirements = raw
+            .requirements
+            .as_ref()
+            .map(|req| Requirements::from_stage1(req, &jinja))
+            .transpose()?
+            .unwrap_or_default();
+
+        let build = Build::from_stage1(&raw.build, &jinja)?;
+        let test = Test::from_stage1(&raw.test, &jinja)?;
+
+        Ok(Self {
+            package,
+            source,
+            build,
+            requirements,
+            test,
+            about,
+            extra: (),
+        })
+    }
+
+    /// Get the package information.
+    pub const fn package(&self) -> &Package {
+        &self.package
+    }
+
+    /// Get the source information.
+    pub fn sources(&self) -> &[Source] {
+        self.source.as_slice()
+    }
+
+    /// Get the build information.
+    pub const fn build(&self) -> &Build {
+        &self.build
+    }
+
+    /// Get the requirements information.
+    pub const fn requirements(&self) -> &Requirements {
+        &self.requirements
+    }
+
+    /// Get the test information.
+    pub const fn test(&self) -> &Test {
+        &self.test
+    }
+
+    /// Get the about information.
+    pub const fn about(&self) -> &About {
+        &self.about
+    }
+}
```

Comment:
I did that because the crate is a library and a binary, so I followed Rust conventions for public API.

This is more on the personal side, but I also felt that using the accessors already returning references improved a bit of the readability of the code that used them